src/promnesia/__main__.py

```diff
@@ -29,8 +42,16 @@ def iter_all_visits() -> Iterator[Res[DbVisit]]:
 
     hook = cfg.hook
 
-    indexers = cfg.sources
-    for idx in indexers:
+    indexers = list(cfg.sources)
+
+    if sources_subset:
+        indexers = _decide_indexers(indexers, sources_subset)
+    logger.info("sources to extract: %s", ", ".join(i.name for i in indexers))
```

Comment:
It can be an Exception (e.g. if it's a 'lazy' indexer), in which case `.name` would cause everything to crash. If you run tox -e mypy-core`, it flags it.

IMO best to keep it iterative, without forcing to list, and move the check below `instance(idx, Exception)` check. Worst case the user will have the subset they specified + a bunch of exceptions.

Reply:
Fixed in 0775abb and tests now pass, except those requiring `private` module.

What is this??

Reply:
Right, checked the 'import private' things -- these are automation scripts in [demos.py](https://github.com/karlicoss/promnesia/blob/master/tests/demos.py ) I used to record demos in [readme](https://github.com/karlicoss/promnesia#readme). They are also serve as kind of end to end tests, however . The private thing was because they imported some cookies which I didn't want to commit in the main repository.
Either way they are not running on github because of this: https://github.com/karlicoss/promnesia/blob/e44a2bda053cc93bb8dc1b84fd1a4461aa5d914f/tox.ini#L23

The `not with_browser` thing excludes tests marked with `with_browser` (so all tests in `demos.py`).
What was the command you used to run? Either way, sorry it's confusing, I guess a better option would be to make the default to run the same what CI is running, and only if user want, explicitly make it run these GUI tests.

There is also another test guilty of this: https://github.com/karlicoss/promnesia/blob/e44a2bda053cc93bb8dc1b84fd1a4461aa5d914f/tests/indexer_test.py#L158-L159 -- it doesn't really run on CI, because it has `CI` env variable set, but it might not be the case for people.

In the meantime, the most reliable command to run is `CI=true tox -e tests`, and you can also restrict them to specific test like this: `CI=true tox -e tests -- -k test`. I'll try to clean up the CI config in the meantime so it's easier for people to contribute.

Reply:
Similar for mypy: `CI=true tox -e mypy-core` -- also has a couple of failures

Reply:
Ok, hopefully i fixed the bug in `do-demo()` and mypy issues, but
i'm `tests/cli.py::test_demo` still fails with this:

```
do_demo( index_as=getattr(args, 'as'), params=args.params, port=args.port, config_file=args.config, name=args.name, sources_subset=args.sources)
*** FileNotFoundError: [Errno 2] No such file or directory: '~/.config/promnesia-test/config.py'
Traceback (most recent call last):
  File "./promnesia.git/src/promnesia/__main__.py", line 145, in do_demo
    config.load_from(config_file)
  File "./promnesia.git/src/promnesia/config.py", line 102, in load_from
    instance = import_config(config_file)
  File "./promnesia.git/src/promnesia/config.py", line 118, in import_config
    spec.loader.exec_module(mod) # type: ignore
  File "<frozen importlib._bootstrap_external>", line 786, in exec_module
  File "<frozen importlib._bootstrap_external>", line 922, in get_code
  File "<frozen importlib._bootstrap_external>", line 979, in get_data
```

Any idea how ` ~/.config/promnesia-test/config.py` file came to be?

Reply:
Also i would like some recommendations on how to structure the test-cases for the new CLI features (the 2 new options), like, which files to create, and what to patch, etc.

And if i get it right, there were no new test-cases for the  "update db' functionality you did in #20, correct @karlicoss.  would t be possible for you to do it?

Reply:
> Any idea how ~/.config/promnesia-test/config.py file came to be?

Yeah, commented about the test https://github.com/karlicoss/promnesia/pull/211#discussion_r588912230

> Also i would like some recommendations on how to structure the test-cases for the new CLI features 

Sure! For tests 'structure', it's currently:

- `tests/indexer_test`  -- usually tests sources and how they extract visits
- `tests/integration_test` -- usually tests indexing, and does actual database checks
- `tests/server_test` -- usually tests database responses
- `tests/cli` -- tests cli interation

I say 'usually' because it's a little messy and the separation is not 100% clear at times. For example, most tests in `server_test` also usually run indexing and extraction, they aren't very 'unit' tests. Not against it, but just with restricted resources full-scale tests give better coverage, at the expense of a bit more confusion when they fail.

So, I only recently started CLI tests in `cli.py` -- and it's actually the one failing now ;) In general would be nice to have more tests there. I think the most important thing is so things don't crash there -- the actual code in `__main__` is fairly covered by other tests.

So maybe adding a couple of tests which do `check_call(promnesia_bin(command_string.split()))` or something like that wouldn't hurt. Probably the most important candidates are commands already mentioned in README and other docs -- would be the worst to break them. 

If any command needs data source, can use `ox_hugo_data` which is already in this file, or `hypothesis_data` which is used in `integration_test`. Note that you might need to run `git submodule update --init --recursive` -- the test data is in submodules (to avoid bloating the main repository). Let me know if you have any issues with this!

> And if i get it right, there were no new test-cases for the "update db' functionality you did 

For tests, at first I thought that I didn't, but it seems there is one! 
https://github.com/karlicoss/promnesia/blob/master/tests/server_test.py#L178
However it's not actually testing anything meaningful, just checks nothing crashes. 

In fact now it would be even less meaningful, because of (unfortunately) hardcoded env variable.
https://github.com/karlicoss/promnesia/blob/master/tests/integration_test.py#L14-L24

I don't mind adding it after we merge though -- I could share a PR with you so you could see what I change and ask questions to make yourself more familiar. Sounds good?

Reply:
Will do my study, and come back,
or you may take over?

Reply:
I think I found it! Was missing the directory, and sql seemed very obscure about it
https://github.com/karlicoss/promnesia/commit/f47c03e63582adb70f18b948d99f3fc92f4dcf54
I could merge and then you rebase, or might be just easier if you cherry pick the commit? 

Curious why it didn't happen before, but probably a different test order or something like that? (sometimes it depends on who knows what). Either way probably not worth investigating too much -- in my experience empty dir never hurts anyway, and eases a lot of pain like that.

If you're curious how -- there is a thing in the github actions pipeline one can uncomment https://github.com/karlicoss/promnesia/blob/master/.github/workflows/main.yml#L36-L37
after that it's possible to SSH onto the machine tests are running on and reproduce

Reply:
Can cherry-pick (have not yet undestand what is it about).

Reply:
All looks good! Happy to merge?

Reply:
Not yet, need to retrofit the old test-case with the env-var.

Reply:
OK, done, please merge.

(i noticed that GitHub run just 1 check o the PR, did you change something in the actions?

Reply:
ignore my last comment about missing GH-actions.

Reply:
No, you were right! I have to manually push it at the moment to trigger.
I experimented and think I figured out how to set it up properly so it works on forked repos/PRs as well, I'll do it later & update the instructions.

Reply:
> IMO best to keep it iterative, without forcing to list, and move the check below instance(idx, Exception) check. 
> Worst case the user will have the subset they specified + a bunch of exceptions.

A note from posterity: i should have kept the case where `--sources` is given as a `list`, as originally implemented by `_decide_indexers()`) - not really gaining much memory by being iterative, sources are not gonna be innumerable.  And would avoid seeing a bunch of irrelevant lines everytime i update a single source.

See also discussion [about iterables in zulip](https://memex.zulipchat.com/#narrow/stream/279601-hpi/topic/Iterator.20vs.20Iterable/near/231798316).